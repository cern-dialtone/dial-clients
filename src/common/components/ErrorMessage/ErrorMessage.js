import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Message } from 'semantic-ui-react';

/**
 * Displays one or several error messages.
 * @param {errors} Array of errors with message (string) and statusCode (string)
 * @param {solutions} Array of string corresponding to the solutions to the error
 */
function ErrorMessage({ errors, solutions }) {
  const [showError, setShowError] = useState(1);
  const results = errors.filter(error => error && error.statusCode);

  // No errors or the user clicked on the close button
  if (results.length < 1 || !showError) {
    return null;
  }

  // Format the errors to display the message and the status code
  const messages = results.map(
    element => `${element.message} (${element.statusCode})`
  );

  /*
   If we have only one error, we display it as header, adding possible solutins as children.

   If we have more than one errors, we only display them, without solutions.
  */
  return (
    <Message
      error
      onDismiss={() => setShowError(!showError)}
      header={
        results.length === 1
          ? `${results[0].message} (${results[0].statusCode})`
          : `An error has occurred`
      }
      list={results.length === 1 ? [...solutions] : [...messages]}
    />
  );
}

ErrorMessage.propTypes = {
  solutions: PropTypes.arrayOf(PropTypes.string.isRequired),
  errors: PropTypes.arrayOf(
    PropTypes.shape({
      message: PropTypes.string.isRequired,
      statusCode: PropTypes.string.isRequired
    })
  )
};

ErrorMessage.defaultProps = {
  solutions: [],
  errors: []
};

export default ErrorMessage;
