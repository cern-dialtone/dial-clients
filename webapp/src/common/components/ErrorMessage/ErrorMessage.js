import React, { Component } from "react";
import PropTypes from "prop-types";
import { translate } from "react-i18next";
import { Message, Segment } from "semantic-ui-react";
import { logMessage } from "common/utils/logs";

export class ErrorMessage extends Component {
  static propTypes = {
    errors: PropTypes.array,
    t: PropTypes.func.isRequired
  };

  styles = {
    margin: 0,
    padding: "0 0 1em 0"
  };

  render() {
    logMessage(`Displaying errors`);
    const { errors } = this.props;

    let results = errors.filter(error => error && error.statusCode);

    if (results.length < 1) {
      return "";
    }
    return (
      <Segment basic style={this.styles}>
        <Message color="red">
          <ul>
            {results.map((element, index) => {
              return (
                <li key={`message${index}`}>
                  {element.message} ({element.statusCode})
                </li>
              );
            })}
          </ul>
        </Message>
      </Segment>
    );
  }
}

export default translate("calls")(ErrorMessage);
