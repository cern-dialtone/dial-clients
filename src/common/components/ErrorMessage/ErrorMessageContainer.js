import { connect } from 'react-redux';
import withPhoneService from 'calls/providers/PhoneProvider/PhoneService';
import ErrorMessage from 'common/components/ErrorMessage/ErrorMessage';

function mapStateToProps({ call, connection, numbers, auth }) {
  return {
    errors: [call.error, connection.error, numbers.error, auth.error]
  };
}

export const ConnectedErrorMessageContainer = connect(
  mapStateToProps,
  null
)(ErrorMessage);

export default withPhoneService(ConnectedErrorMessageContainer);
