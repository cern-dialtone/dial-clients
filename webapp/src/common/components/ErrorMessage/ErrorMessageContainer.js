import { connect } from "react-redux";
import { phoneService } from "calls/providers/PhoneProvider/PhoneProvider";
import ErrorMessage from "common/components/ErrorMessage/ErrorMessage";

function mapStateToProps({ calls, auth }) {
  return {
    errors: [
      calls.call.error,
      calls.connection.error,
      calls.numbers.error,
      auth.error
    ]
  };
}

export const ErrorMessageContainer = connect(
  mapStateToProps,
  null
)(ErrorMessage);

export default phoneService(ErrorMessageContainer);
