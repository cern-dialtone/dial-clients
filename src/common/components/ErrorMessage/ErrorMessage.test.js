import React from 'react';
import { render, waitForElement, fireEvent } from '@testing-library/react';
import ErrorMessage from './ErrorMessage';

global.remote = {
  require: jest.fn()
};

describe('ErrorMessage tests', () => {
  it('renders component and includes one error', async () => {
    const errors = [{ message: 'This is error 1', statusCode: '1' }];
    const { getByText, asFragment } = render(<ErrorMessage errors={errors} />);

    await waitForElement(() => getByText('This is error 1 (1)'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders null without any errors', async () => {
    const { container } = render(<ErrorMessage />);
    expect(container.firstChild).toBeNull();
  });

  it('renders component and includes one error and solutions', async () => {
    const errors = [{ message: 'This is error 1', statusCode: '1' }];
    const solutions = [
      'This is solution 1',
      'This is solution 2',
      'This is solution 3'
    ];
    const { getByText, asFragment } = render(
      <ErrorMessage errors={errors} solutions={solutions} />
    );

    await waitForElement(() => getByText('This is error 1 (1)'));
    await waitForElement(() => getByText('This is solution 1'));
    await waitForElement(() => getByText('This is solution 2'));
    await waitForElement(() => getByText('This is solution 3'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders component and includes multiple errors', async () => {
    const errors = [
      { message: 'This is error 1', statusCode: '1' },
      { message: 'This is error 2', statusCode: '2' }
    ];
    const { getByText, asFragment } = render(<ErrorMessage errors={errors} />);

    await waitForElement(() => getByText('An error has occurred'));
    await waitForElement(() => getByText('This is error 1 (1)'));
    await waitForElement(() => getByText('This is error 2 (2)'));

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders multiple errors without solutions', async () => {
    const errors = [
      { message: 'This is error 1', statusCode: '1' },
      { message: 'This is error 2', statusCode: '2' }
    ];
    const solutions = [
      'This is solution 1',
      'This is solution 2',
      'This is solution 3'
    ];
    const { getByText, asFragment, queryByText } = render(
      <ErrorMessage errors={errors} solutions={solutions} />
    );

    await waitForElement(() => getByText('This is error 1 (1)'));

    const solution1Text = queryByText('This is solution 1');
    expect(solution1Text).toBeNull(); // it doesn't exist
    const solution2Text = queryByText('This is solution 2');
    expect(solution2Text).toBeNull(); // it doesn't exist
    const solution3Text = queryByText('This is solution 3');
    expect(solution3Text).toBeNull(); // it doesn't exist

    expect(asFragment()).toMatchSnapshot();
  });

  it('handles click on close button', async () => {
    const errors = [{ message: 'This is error 1', statusCode: '1' }];
    const { getByText, container } = render(<ErrorMessage errors={errors} />);
    await waitForElement(() => getByText('This is error 1 (1)'));
    expect(container.firstChild).not.toBeNull();
    // When clicked on the close button, the component will return null
    const foo = container.querySelector('i');
    fireEvent.click(foo);
    expect(container.firstChild).toBeNull();
  });
});
