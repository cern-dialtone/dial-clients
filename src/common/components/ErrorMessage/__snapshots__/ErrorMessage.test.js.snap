// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage tests renders component and includes multiple errors 1`] = `
<DocumentFragment>
  <div
    class="ui error message"
  >
    <i
      aria-hidden="true"
      class="close icon"
    />
    <div
      class="content"
    >
      <div
        class="header"
      >
        An error has occurred
      </div>
      <ul
        class="list"
      >
        <li
          class="content"
        >
          This is error 1 (1)
        </li>
        <li
          class="content"
        >
          This is error 2 (2)
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ErrorMessage tests renders component and includes one error 1`] = `
<DocumentFragment>
  <div
    class="ui error message"
  >
    <i
      aria-hidden="true"
      class="close icon"
    />
    <div
      class="content"
    >
      <div
        class="header"
      >
        This is error 1 (1)
      </div>
      <ul
        class="list"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`ErrorMessage tests renders component and includes one error and solutions 1`] = `
<DocumentFragment>
  <div
    class="ui error message"
  >
    <i
      aria-hidden="true"
      class="close icon"
    />
    <div
      class="content"
    >
      <div
        class="header"
      >
        This is error 1 (1)
      </div>
      <ul
        class="list"
      >
        <li
          class="content"
        >
          This is solution 1
        </li>
        <li
          class="content"
        >
          This is solution 2
        </li>
        <li
          class="content"
        >
          This is solution 3
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ErrorMessage tests renders multiple errors without solutions 1`] = `
<DocumentFragment>
  <div
    class="ui error message"
  >
    <i
      aria-hidden="true"
      class="close icon"
    />
    <div
      class="content"
    >
      <div
        class="header"
      >
        An error has occurred
      </div>
      <ul
        class="list"
      >
        <li
          class="content"
        >
          This is error 1 (1)
        </li>
        <li
          class="content"
        >
          This is error 2 (2)
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
